diff --git a/compat/mingw.c b/compat/mingw.c
--- a/compat/mingw.c
+++ b/compat/mingw.c
@@ -1745,9 +1745,13 @@ char *mingw_getcwd(char *pointer, int len)
 	}
 	hnd = CreateFileW(cwd, 0,
 			  FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
 			  OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
 	if (hnd != INVALID_HANDLE_VALUE) {
 		ret = GetFinalPathNameByHandleW(hnd, wpointer, ARRAY_SIZE(wpointer), 0);
+		if (!ret && GetLastError() == ERROR_ACCESS_DENIED)
+			ret = GetFinalPathNameByHandleW(hnd, wpointer,
+						ARRAY_SIZE(wpointer),
+						FILE_NAME_OPENED | VOLUME_NAME_DOS);
 		CloseHandle(hnd);
 		if (!ret || ret >= ARRAY_SIZE(wpointer)) {
 			ret = GetLongPathNameW(cwd, wpointer, ARRAY_SIZE(wpointer));
